api/views.py

```diff
@@ -0,0 +1,176 @@
+        employees = Employee.objects.all()
+        sum_of_salaries = 0
+        for employee in employees:
+            sum_of_salaries += int(employee.salary)
+        return Response("Summary cost of all salaries: " + str(sum_of_salaries))
+
+
+class IncomeViewSet(viewsets.ModelViewSet):
+    queryset = Income.objects.all()
+    serializer_class = IncomeSerializer
+    filter_fields = ('name', 'date')
+
+    @action(detail=False, methods=['get'])
+    def last_incomes(self, request, *args, **kwargs):
+        number_of_incomes_to_show = request.GET.get('number_of_incomes', None)
+        if number_of_incomes_to_show is not None:
+            incomes = Income.objects.order_by('date')[:int(number_of_incomes_to_show)]
+        else:
+            incomes = Income.objects.order_by('date')
+        serializer = EventSerializer(incomes, many=True)
+        return Response(serializer.data)
+
+    @action(detail=False, methods=['get'])
+    def sum_of_cost(self, request, *args, **kwargs):
+        days_back = request.GET.get('days', 7)
+        current_date = datetime.today().date()
+        date_for_filter = current_date - timedelta(days=int(days_back))
+        incomes = Income.objects.filter(date__gte=date_for_filter)
+        sum = 0
+        for income in incomes:
+            sum += int(income.sum)
+        return Response("Total income from the last " + str(days_back) + " days is: " + str(sum))
+
+
+class OutcomeViewSet(viewsets.ModelViewSet):
+    queryset = Outcome.objects.all()
+    serializer_class = OutcomeSerializer
+    filter_fields = ('name', 'date')
+
+    @action(detail=False, methods=['get'])
+    def last_outcomes(self, request, *args, **kwargs):
+        number_of_outcomes_to_show = request.GET.get('number_of_outcomes', None)
+        if number_of_outcomes_to_show is not None:
+            outcomes = Outcome.objects.order_by('date')[:int(number_of_outcomes_to_show)]
+        else:
+            outcomes = Outcome.objects.order_by('date')
+        serializer = EventSerializer(outcomes, many=True)
+        return Response(serializer.data)
+
+    @action(detail=False, methods=['get'])
+    def sum_of_cost(self, request, *args, **kwargs):
+        days_back = request.GET.get('days', 7)
+        current_date = datetime.today().date()
+        date_for_filter = current_date - timedelta(days=int(days_back))
+        outcomes = Outcome.objects.filter(date__gte=date_for_filter)
+        sum = 0
+        for outcome in outcomes:
+            sum += int(outcome.sum)
+        return Response("Total outcome from the last " + str(days_back) + " days is: " + str(sum))
+
+
+class EventViewSet(viewsets.ModelViewSet):
+    queryset = Event.objects.all()
+    serializer_class = EventMiniSerializer
+
+    def get_future_events(self):
+        events = Event.objects.filter(date__gte=datetime.today().date())
+        current_date = datetime.today().date()
+        current_time = datetime.today().time()
+        for event in events:
```

Comment:
Yet again, imagine we have millions or more events in the database and there are maaany of them for each day, and we're looping over them all the time. This can be done on a query level, in one line:

```
from django.db.models import Q

current_date = datetime.today().date()
current_time = datetime.today().time()
return Event.objects.filter(Q(date__gte=current_date) & Q(time__lte=current_time))
```

If I'm not mistaken, we cannot use simple `Event.objects.filter(date__gte=current_date, time__lte=current_time)`, because this would give us logical `OR` in the place of the comma. Using `Q` allows us to make more advanced filtering and use all the logical operators.

Also note, that it would most probably be better to have one `datetime` field on the event. It's easy to grab either the date or time part from this field and you can then simplify things like filtering and sorting. Now it'll always require to filter on two fields.

Reply:
ok